### PR TITLE
Bug 1877693: dnsrecords specify recordTTL as 30 but the value is nul…

### PR DESCRIPTION
**Pull request by candita**
Bug 1877693 - dnsrecords specify recordTTL as 30 but the value is null in AWS Route 53.  Per AWS API contract, Alias targets may not contain TTL, only CNAME may contain TTL.  See https://github.com/openshift/cluster-ingress-operator/pull/454#discussion_r483898974.  Documenting this behavior.

### CODE REVIEW

File: vendor/github.com/openshift/api/operatoringress/v1/types.go [hunk 1]
```diff
@@ -45,6 +45,8 @@ type DNSRecordSpec struct {
 	// +required
 	RecordType DNSRecordType `json:"recordType"`
 	// recordTTL is the record TTL in seconds. If zero, the default is 30.
+	// RecordTTL will not be used in AWS regions Alias targets, but
```
Comment (sgreene570) [inline vendor/github.com/openshift/api/operatoringress/v1/types.go:48]: The API definitions are vendored from `openshift/api`. For the changes to this file to take affect, you would want to make these same changes in the `openshift/api` github repo (specifically here https://github.com/openshift/api/blob/master/operatoringress/v1/types.go). Then, when we update the vendored version of `openshift/api` in `go.mod`, your changes would propagate into the ingress operator's repo :smile:

File: pkg/dns/aws/dns.go [hunk 2]
```diff
@@ -409,6 +409,7 @@ func (m *Provider) change(record *iov1.DNSRecord, zone configv1.DNSZone, action
 // target in targetHostedZoneID. An Alias record type is used for all regions
 // other than GovCloud (CNAME). See the following for additional details:
 // https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/govcloud-r53.html
+// Note that by API contract, TTL cannot be specified for an AliasTarget
```
Comment (sgreene570) [inline pkg/dns/aws/dns.go:412]:
```suggestion
// Note that by API contract, TTL cannot be specified for an AliasTarget.
```

**Comment by sgreene570** [comment]
By the way, the bugzilla PR automation expects the PR title to contain the following substring:
`Bug xxxxx:` Instead of `Bug xxxxx -`

File: vendor/github.com/openshift/api/operatoringress/v1/types.go (hunk 1 above)
Comment (candita) [reply vendor/github.com/openshift/api/operatoringress/v1/types.go:48]: Thanks Steve.  I noticed this in the CI output yesterday and created [a PR](https://github.com/openshift/api/pull/786) in openshift/api after consulting with Miciah.

**Comment by Miciah** [comment]
There should be changes under `vendor/github.com/openshift/api/`.  Try running `GO111MODULE=on GOFLAGS=-mod=vendor go mod vendor`.

**Comment by sgreene570** [comment]
@Miciah I think the `openshift/api` changes introduce unit test breakages resultant of https://github.com/openshift/api/pull/776

**Comment by Miciah** [comment]
> @Miciah I think the `openshift/api` bump introduces unit test breakages resultant of [openshift/api#776](https://github.com/openshift/api/pull/776)

Indeed.

/hold

Candace, do you want to add a commit to fix the unit tests?  If you prefer, I can make a separate PR to fix the unit tests, and then you can rebase this PR.

**Comment by candita** [comment]
@Miciah sure, I will add the unit test changes.

File: pkg/operator/controller/ingress/deployment_test.go [hunk 3]
```diff
@@ -262,7 +262,7 @@ func TestDesiredRouterDeployment(t *testing.T) {
 			},
 			HttpLogFormat: "%ci:%cp [%t] %ft %b/%s %B %bq %HM %HU %HV",
 			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{
-				{MatchType: "Prefix", NamePrefix: "foo"},
+				0: {IngressControllerCaptureHTTPCookieUnion: operatorv1.IngressControllerCaptureHTTPCookieUnion{MatchType: "Prefix", NamePrefix: "foo"}},
 			},
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:266]: You can omit the index.  Maybe something like this:  
```suggestion
			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{{
				IngressControllerCaptureHTTPCookieUnion: operatorv1.IngressControllerCaptureHTTPCookieUnion{
					MatchType:  "Prefix",
					NamePrefix: "foo",
				},
			}},
```

Comment (candita) [reply pkg/operator/controller/ingress/deployment_test.go]: Ok, reformatted it too.

File: pkg/operator/controller/ingress/deployment_test.go [hunk 4]
```diff
@@ -261,9 +261,11 @@ func TestDesiredRouterDeployment(t *testing.T) {
 				Container: &operatorv1.ContainerLoggingDestinationParameters{},
 			},
 			HttpLogFormat: "%ci:%cp [%t] %ft %b/%s %B %bq %HM %HU %HV",
-			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{
-				{MatchType: "Prefix", NamePrefix: "foo"},
-			},
+			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{{
+				IngressControllerCaptureHTTPCookieUnion: operatorv1.IngressControllerCaptureHTTPCookieUnion{
+					MatchType:  "Prefix",
+					NamePrefix: "foo"},
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:267]:
```suggestion
					NamePrefix: "foo",
				},
```

File: pkg/operator/controller/ingress/deployment_test.go [hunk 5]
```diff
@@ -397,9 +399,12 @@ func TestDesiredRouterDeployment(t *testing.T) {
 					{Name: "Location", MaxLength: 15},
 				},
 			},
-			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{
-				{MatchType: "Exact", Name: "foo", MaxLength: 15},
-			},
+			HTTPCaptureCookies: []operatorv1.IngressControllerCaptureHTTPCookie{{
+				IngressControllerCaptureHTTPCookieUnion: operatorv1.IngressControllerCaptureHTTPCookieUnion{
+					MatchType: "Exact",
+					Name:      "foo"},
```
Comment (Miciah) [inline pkg/operator/controller/ingress/deployment_test.go:405]:
```suggestion
					Name:      "foo",
				},
```

**Review by Miciah — COMMENTED** [review]
Tiny formatting issue.

**Comment by Miciah** [comment]
Oh, doesn't this need to be fixed too?  
https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L1443-L1445

File: pkg/operator/controller/ingress/deployment_test.go (hunk 4 above)
Comment (candita) [reply pkg/operator/controller/ingress/deployment_test.go]: Tiny fixed!

**Comment by candita** [comment]
@Miciah  thanks for the pointers - incorporated all.

**Comment by Miciah** [comment]
Can you rebase and rerun `make generate`?  (We generally don't use merge commits in PRs.)

**Comment by candita** [comment]
> Can you rebase and rerun `make generate`? (We generally don't use merge commits in PRs.)

Will do.

**Pull request merged**